src/passes/Precompute.cpp

```diff
@@ -240,6 +240,38 @@ class PrecomputingExpressionRunner
     // string.encode_wtf16_array anyhow.)
     return Flow(NONCONSTANT_FLOW);
   }
+
+  Flow visitStringSliceWTF(StringSliceWTF* curr) {
+    auto flow = Super::visitStringSliceWTF(curr);
```

Comment:
I think a comment like I suggested here is useful, to explain why we override here at all.

I am not sure my text there is right, since I'm not sure I follow the reason here. Is the problem the slicing, or putting it back into a constant?

If it is just wrong to slice, then this looks right as it is.

If the problem is putting a split surrogate into a `string.const`, then we can enforce that at the `string.const` itself? The place for that would be `bool canEmitConstantFor(const Literal& value)` lower down in this file - it can return false for invalid string constants (no matter how they were created - that would handle not just slicing but concatenation too). But, if it is simpler to do it here with an explanation, sgtm too.

Reply:
The problem is ending up with a `string.const` that does not represent a valid UTF-16. So it would be ok to do something like `sting.measure` to the slice that is not UTF and that would be fine, but this patch would prevent it.

`canEmitConstantFor` seems like the proper place. Let me test it and will send a pr with that.